File: PanFlow.API/PanFlow.API/Features/Analysis/Services/AnalysisService.cs
using Microsoft.EntityFrameworkCore;
using PanFlow.API.Data;
using PanFlow.API.Features.Analysis.DTOs.Read;
using PanFlow.API.Shared.DTOs;

namespace PanFlow.API.Features.Analysis.Services
{
    public class AnalysisService : IAnalysisService
    {
        private readonly AppDbContext _context;

        public AnalysisService(AppDbContext context)
        {
            _context = context;
        }




        public async Task<GeneralResponseDto<ReadAnalysisResponse>> GetAnalysis(string userId)
        {
            // Cheack there is user Id 
            if (string.IsNullOrWhiteSpace(userId))
            {
                return GeneralResponseDto<ReadAnalysisResponse>.Failure("there is no User Id");
            }
            var user = await _context.Users.FindAsync(userId);
            if (user == null)
            {
                return GeneralResponseDto<ReadAnalysisResponse>.Failure("User Is Not Esixt");
            }

            // ── Days ──
            // كل أيام المستخدم (غير المحذوفة) مع الـ DayHabits بتاعتها
            var days = await _context.Days
                .Where(d => d.UserId == userId && d.IsDeleted == false)
                .Include(d => d.DayHabits)
                .OrderBy(d => d.DayDate)
                .ToListAsync();

            var daysAnalytics = days.Select(d => new DayAnalyticsResponse
            {
                DayId = d.DayId,
                Date = d.DayDate,
                TotalHabits = d.DayHabits.Count,
                CompletedHabits = d.DayHabits.Count(dh => dh.IsChecked),
                CompletionPercentage = d.DayHabits.Count == 0
                    ? 0
                    : Math.Round((double)d.DayHabits.Count(dh => dh.IsChecked) / d.DayHabits.Count * 100, 2)
            }).ToList();

            // ── Habits ──
            // كل عادات المستخدم النشطة (من خلال الـ Aspects بتاعته)
            var habits = await _context.Habits
                .Where(h => h.Aspect.UserId == userId && h.IsDeleted == false)
                .Include(h => h.Aspect)
                .ToListAsync();

            var dayIds = days.Select(d => d.DayId).ToList();

            // كل الـ DayHabits اللي بتخص أيام المستخدم دي (مرة واحدة عشان منعملش استعلام لكل عادة)
            var dayHabitsForUser = await _context.DayHabits
                .Where(dh => dayIds.Contains(dh.DayId))
                .ToListAsync();

            var habitsAnalytics = habits.Select(h =>
            {
                var relatedDayHabits = dayHabitsForUser.Where(dh => dh.HabitId == h.HabitId).ToList();

                var totalDays = relatedDayHabits.Count;
                var completedDays = relatedDayHabits.Count(dh => dh.IsChecked);

                return new HabitAnalyticsResponse
                {
                    HabitId = h.HabitId,
                    HabitName = h.HabitName,
                    HabitColor = h.Aspect.AspectColor,
                    TotalDays = totalDays,
                    CompletedDays = completedDays,
                    CompletionRate = totalDays == 0
                        ? 0
                        : Math.Round((double)completedDays / totalDays * 100, 2),
                    HasData = totalDays > 0
                };
            }).ToList();

            var response = new ReadAnalysisResponse
            {
                Days = daysAnalytics,
                Habits = habitsAnalytics
            };

            return GeneralResponseDto<ReadAnalysisResponse>.Success("Analysis fetched successfully", response);
        }
    }
}

File: PanFlow.UI/src/app/features/analysis/components/analysis/analysis-service.ts
import { Injectable, inject, signal, computed } from '@angular/core';
import { AnalysisAPI } from '../../services/analysis-API';

export interface DayAnalytics {
  dayId: string;
  date: string;
  completionPercentage: number;
  totalHabits: number;
  completedHabits: number;
}

export interface HabitAnalytics {
  habitId: string;
  habitName: string;
  completionRate: number;
  totalDays: number;
  completedDays: number;
  habitColor: string;
  hasData: boolean;
}

@Injectable()
export class AnalysisService {
  private analysisAPI = inject(AnalysisAPI);

  // States
  isLoading = signal<boolean>(false);
  allDays = signal<DayAnalytics[]>([]);
  allHabits = signal<HabitAnalytics[]>([]);

  // Computed values
  bestDay = computed(() => {
    const days = this.allDays();
    return days.length > 0 ? days.reduce((a, b) => a.completionPercentage > b.completionPercentage ? a : b) : null;
  });

  worstDay = computed(() => {
    const days = this.allDays();
    return days.length > 0 ? days.reduce((a, b) => a.completionPercentage < b.completionPercentage ? a : b) : null;
  });

  // بنستبعد العادات اللي لسه معملتلهاش أي Day (hasData: false) من المقارنة
  // عشان منقولش "أسوأ عادة" على عادة لسه ما اتتابعتش أصلاً
  trackedHabits = computed(() => this.allHabits().filter(h => h.hasData));

  bestHabit = computed(() => {
    const habits = this.trackedHabits();
    return habits.length > 0 ? habits.reduce((a, b) => a.completionRate > b.completionRate ? a : b) : null;
  });

  worstHabit = computed(() => {
    const habits = this.trackedHabits();
    return habits.length > 0 ? habits.reduce((a, b) => a.completionRate < b.completionRate ? a : b) : null;
  });

  averageCompletion = computed(() => {
    const days = this.allDays();
    if (days.length === 0) return 0;
    const sum = days.reduce((acc, day) => acc + day.completionPercentage, 0);
    return Math.round(sum / days.length);
  });

  totalDaysCount = computed(() => this.allDays().length);

  /**
   * الحساب كله بقى شغال في الـ Backend (Analysis Feature)
   * الـ endpoint الواحد ده بيرجع الأيام والعادات جاهزين بالنسب الصحيحة،
   * فاختفت مشكلة إن completionRate بتاع العادات كان بيتحط صفر دايمًا.
   */
  loadAnalytics() {
    this.isLoading.set(true);

    this.analysisAPI.read().subscribe({
      next: (response) => {
        const days = response.data?.days ?? [];
        const habits = response.data?.habits ?? [];

        const daysAnalytics: DayAnalytics[] = days.map(day => ({
          dayId: day.dayId,
          date: day.date,
          completionPercentage: day.completionPercentage ?? 0,
          totalHabits: day.totalHabits ?? 0,
          completedHabits: day.completedHabits ?? 0
        }));
        this.allDays.set(daysAnalytics);

        const habitsAnalytics: HabitAnalytics[] = habits.map(habit => ({
          habitId: habit.habitId,
          habitName: habit.habitName,
          completionRate: habit.completionRate ?? 0,
          totalDays: habit.totalDays ?? 0,
          completedDays: habit.completedDays ?? 0,
          habitColor: habit.habitColor ?? '#cc8500',
          hasData: habit.hasData ?? (habit.totalDays ?? 0) > 0
        }));
        this.allHabits.set(habitsAnalytics);

        this.isLoading.set(false);
      },
      error: () => {
        this.isLoading.set(false);
      }
    });
  }
}
